rafthttp/pipeline_test.go

```diff
@@ -57,24 +57,27 @@ func TestPipelineSend(t *testing.T) {
 // TestPipelineKeepSendingWhenPostError tests that pipeline can keep
 // sending messages if previous messages meet post error.
 func TestPipelineKeepSendingWhenPostError(t *testing.T) {
-	tr := &respRoundTripper{err: fmt.Errorf("roundtrip error")}
+	tr := &respRoundTripper{rec: testutil.NewRecorderStream(), err: fmt.Errorf("roundtrip error")}
 	picker := mustNewURLPicker(t, []string{"http://localhost:2380"})
 	fs := &stats.FollowerStats{}
 	tp := &Transport{pipelineRt: tr}
 	p := newPipeline(tp, picker, types.ID(2), types.ID(1), types.ID(1), newPeerStatus(types.ID(1)), fs, &fakeRaft{}, nil)
+	defer p.stop()
 
 	for i := 0; i < 50; i++ {
 		p.msgc <- raftpb.Message{Type: raftpb.MsgApp}
 	}
-	testutil.WaitSchedule()
-	p.stop()
 
-	// check it send out 50 requests
-	tr.mu.Lock()
-	defer tr.mu.Unlock()
-	if tr.reqCount != 50 {
-		t.Errorf("request count = %d, want 50", tr.reqCount)
-	}
+	done := make(chan struct{})
+
+	go func() {
```

Comment:
The goroutine runs after `msgc` is filled by the for loop (which I assume is where it would block on recorderstream stuff)-- where's the concurrency?